File: Week2/ArrayProblem/RemoveDuplicateSortedInplace.java
class RemoveDuplicateSortedInplace{
/*
 * Given an integer array nums sorted in non-decreasing order, remove the duplicates in-place such that each unique element appears only once. 
 * The relative order of the elements should be kept the same. Then return the number of unique elements in nums.
 * https://leetcode.com/problems/remove-duplicates-from-sorted-array/
 * T(c) -> O(n), S(c) -> O(1)
 */
    static int removeDuplicates(int[] nums) {
        
        /* using 2 pointer approach, i points to unique element, j points to current element */
        int i = 0;
        int j = 1;
        /* we will shift unique elements at the first part of array */
        while(j<nums.length)
        {
            /* incase of non duplicate element j, shift it to next element of unique i */
            if(nums[i] != nums[j])
            {
                nums[i+1] = nums[j];
                i++;
            }
            /* if dupicate, skip current element */
            j++;
        }
        return i+1;
    }

    public static void main(String[] args) {
        int nums[] = {0,0,1,1,1,2,2,3,3,4};
        int k = removeDuplicates(nums);
        for(int i=0;i<k;i++)
        {
            System.out.print(nums[i]+" ");
        }
    }
}

File: Week2/Python/ArrayProblem/RemoveDuplicateSortedInplace.py
"""
Given an integer array nums sorted in non-decreasing order, remove the duplicates in-place such that each unique element 
appears only once. The relative order of the elements should be kept the same. 
Then return the number of unique elements in nums.

https://leetcode.com/problems/remove-duplicates-from-sorted-array/
"""

def removeDuplicates(nums):
    """ T(c) => O(n) , S(c) -> O(1) """
    # using 2 pointer approach
    # track unique element
    i = 0
    # point to current element till end
    j = 1
    
    n = len(nums)
    
    while j<n:
        # if not duplicate, shift current element after ith place
        if nums[i] != nums[j]:
            nums[i+1] = nums[j]
            # check with next unique element
            i += 1
        
        # if duplicate element, bypass duplicate and check for next one
        j += 1
        
    return i+1


arr = [0,0,1,1,1,2,2,3,3,4]

for i in range(len(arr)):
    print(arr[i] , end = " ")

print()

n = removeDuplicates(arr)
for i in range(n):
    print(arr[i] , end = " ")
